numpy/distutils/ccompiler.py

```diff
@@ -491,14 +489,23 @@ def CCompiler_customize(self, dist, need_cxx=0):
     # See FCompiler.customize for suggested usage.
     log.info('customize %s' % (self.__class__.__name__))
     customize_compiler(self)
+
     if need_cxx:
         # In general, distutils uses -Wstrict-prototypes, but this option is
         # not valid for C++ code, only for C.  Remove it if it's there to
         # avoid a spurious warning on every compilation.
-        try:
-            self.compiler_so.remove('-Wstrict-prototypes')
-        except (AttributeError, ValueError):
-            pass
+        #
+        # Distutils don't differentiate between CFLAGS and CPPFLAGS,
+        # so we have to filter out any invalid C++ flags as workaround.
+        #
+        # TODO: Test flags instead of filter them manualy
+        ignore_c_flags = {
+            '-Werror=implicit-function-declaration', '-Wstrict-prototypes'
+        }
+        if hasattr(self, 'compiler_so'):
```

Comment:
Just a guess, but the `-Werror=implicit-function-declaration` flag is present in both `self.compiler_so` and `self.compiler`. It's only removed from the former here. Maybe removing it from both is healthy; the code is too much of a spaghetti to be sure that only `compiler_so` will be used.

Reply:
parm `need_cxx` only used by `build_ext` and `build_clib` that's why this filter has never been called, I took my changes back, kept the original one as-is inside `compiler_cxx()` now we have two C++ filters and we want to put them in one place.

Reply:
I suggest just calling `customize(need_cxx)` inside `cxx_compiler()` which already clones the compiler instance, and we count only on method `cxx_compiler()` in the following code and the other one in `build_clib`  to avoid duplicate initializing and unnecessary copying.

https://github.com/numpy/numpy/blob/20a2a70129e3b29907e2298c0906d9910fbcbd8c/numpy/distutils/command/build_ext.py#L256-L265